protocols/kad/src/behaviour/test.rs

```diff
@@ -1287,3 +1287,87 @@ fn network_behaviour_inject_address_change() {
         kademlia.addresses_of_peer(&remote_peer_id),
     );
 }
+
+#[test]
+fn client_mode() {
+    let mut cfg = KademliaConfig::default();
+    cfg.set_mode(Mode::Client);
+
+    // Create a server and client peer.
+    let client_swarm = build_node_with_config(cfg);
+    let server_swarm = build_node();
+
+    let mut swarms = Vec::new();
+    swarms.push(client_swarm.1);
+    swarms.push(server_swarm.1);
+
+    let mut swarm_addrs = Vec::new();
+    swarm_addrs.push(client_swarm.0);
+    swarm_addrs.push(server_swarm.0);
+
+    let swarm_ids: Vec<_> = swarms
+        .iter()
+        .map(|swarm| swarm.local_peer_id())
+        .cloned()
+        .collect();
+
+    block_on(poll_fn(move |ctx| {
+        // flags to make sure both peers are listening
+        let mut client_listening = false;
+        let mut server_listening = false;
+
+        for swarm in swarms.iter_mut() {
+            loop {
+                match swarm.poll_next_unpin(ctx) {
+                    Poll::Ready(Some(SwarmEvent::NewListenAddr {
+                        listener_id: _,
+                        address,
+                    })) => {
+                        if address == swarm_addrs[0] {
+                            client_listening = true;
+                            return Poll::Ready(());
+                        }
+                        if address == swarm_addrs[1] {
+                            server_listening = true;
+                            return Poll::Ready(());
+                        }
+                    }
+                    Poll::Ready(_) => {
+                        if swarm.local_peer_id() == &swarm_ids[0] && client_listening == true {
+                            let bucket = swarm.behaviour_mut().kbucket(swarm_ids[1].clone());
+                            assert!(bucket.is_some());
+                            // Check if the client peer has the server peer.
+                            assert!(
+                                bucket
+                                    .unwrap()
+                                    .iter()
+                                    .any(|p| p.node.key.preimage() == &swarm_ids[1]),
+                                "The client peer does not have the server node"
+                            );
+
+                            return Poll::Ready(());
+                        }
+
+                        if swarm.local_peer_id() == &swarm_ids[1] && server_listening == true {
+                            let bucket =
+                                swarm.behaviour_mut().kbucket(swarm_ids[0].clone()).unwrap();
```

Comment:
If the `server` peer doesn't have the `client` peer in its routing tables won't `bucket` be `None`?

Reply:
The bucket should not exist in the first place, yes.

Reply:
This is strange, indeed. I wonder how the test is still passing.